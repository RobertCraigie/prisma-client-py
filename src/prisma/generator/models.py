import sys
import enum
import importlib
from pathlib import Path
from keyword import iskeyword
from importlib import machinery, util as importlib_util
from importlib.abc import InspectLoader
from contextvars import ContextVar
from typing import (
    Any,
    Optional,
    List,
    Tuple,
    Union,
    Iterator,
    Dict,
    Type,
    TYPE_CHECKING,
)

import click
from pydantic import (
    BaseModel as PydanticBaseModel,
    BaseSettings,
    Extra,
    Field as FieldInfo,
    validator,
    root_validator,
)

try:
    from pydantic.env_settings import SettingsSourceCallable
except ImportError:
    SettingsSourceCallable = Any  # type: ignore


# NOTE: this does not represent all the data that is passed by prisma

ATOMIC_FIELD_TYPES = ['Int', 'BigInt', 'Float']

TYPE_MAPPING = {
    'String': 'str',
    'DateTime': 'datetime.datetime',
    'Boolean': 'bool',
    'Int': 'int',
    'Float': 'float',
    'BigInt': 'int',
    'Json': '\'fields.Json\'',
}
FILTER_TYPES = [
    'String',
    'DateTime',
    'Boolean',
    'Int',
    'BigInt',
    'Float',
    'Json',
]

data_ctx: ContextVar['Data'] = ContextVar('data_ctx')


def get_config() -> 'Config':
    return data_ctx.get().generator.config


def get_datamodel() -> 'Datamodel':
    return data_ctx.get().dmmf.datamodel


def _module_spec_serializer(spec: machinery.ModuleSpec) -> str:
    assert spec.origin is not None, 'Cannot serialize module with no origin'
    return spec.origin


class BaseModel(PydanticBaseModel):
    class Config:
        json_encoders: Dict[Type[Any], Any] = {
            machinery.ModuleSpec: _module_spec_serializer,
        }


class InterfaceChoices(str, enum.Enum):
    sync = 'sync'
    asyncio = 'asyncio'


class Module(BaseModel):
    spec: machinery.ModuleSpec

    class Config(BaseModel.Config):
        arbitrary_types_allowed: bool = True

    @validator('spec', pre=True, allow_reuse=True)
    @classmethod
    def spec_validator(cls, value: Optional[str]) -> machinery.ModuleSpec:
        spec: Optional[machinery.ModuleSpec] = None

        if value is None:
            value = 'prisma/partial_types.py'

        path = Path.cwd().joinpath(value)
        if path.exists():
            spec = importlib_util.spec_from_file_location(
                'prisma.partial_type_generator', value
            )
        elif value.startswith('.'):
            raise ValueError(
                f'No file found at {value} and relative imports are not allowed.'
            )
        else:
            try:
                spec = importlib_util.find_spec(value)
            except ModuleNotFoundError:
                spec = None

        if spec is None:
            raise ValueError(f'Could not find a python file or module at {value}')

        return spec

    def run(self) -> None:
        importlib.invalidate_caches()
        mod = importlib_util.module_from_spec(self.spec)
        loader = self.spec.loader
        assert loader is not None, 'Expected an import loader to exist.'
        assert isinstance(
            loader, InspectLoader
        ), f'Cannot execute module from loader type: {type(loader)}'

        try:
            loader.exec_module(mod)
        except:
            print('An exception ocurred while running the partial type generator')
            raise


class Data(BaseModel):
    """Root model for the data that prisma provides to the generator."""

    datamodel: str
    version: str
    generator: 'Generator'
    dmmf: 'DMMF' = FieldInfo(alias='dmmf')
    schema_path: str = FieldInfo(alias='schemaPath')
    datasources: List['Datasource'] = FieldInfo(alias='datasources')

    # TODO
    other_generators: List[Any] = FieldInfo(alias='otherGenerators')

    @classmethod
    def parse_obj(cls, obj: Any) -> 'Data':
        data = super().parse_obj(obj)
        data_ctx.set(data)
        return data


class Datasource(BaseModel):
    # TODO: provider enums
    name: str
    provider: str
    active_provider: str = FieldInfo(alias='activeProvider')
    url: 'OptionalValueFromEnvVar'


class Generator(BaseModel):
    name: str
    output: 'ValueFromEnvVar'
    provider: 'ValueFromEnvVar'
    config: 'Config'
    binary_targets: List['ValueFromEnvVar'] = FieldInfo(alias='binaryTargets')
    preview_features: List[str] = FieldInfo(alias='previewFeatures')

    @validator('binary_targets')
    @classmethod
    def warn_binary_targets(
        cls, targets: List['ValueFromEnvVar']
    ) -> List['ValueFromEnvVar']:
        if targets and any(target.value != 'native' for target in targets):
            click.echo(
                click.style(
                    'Warning: '
                    'The binaryTargets option is not currently supported by Prisma Client Python',
                    fg='yellow',
                ),
                file=sys.stdout,
            )

        return targets


class ValueFromEnvVar(BaseModel):
    value: str
    from_env_var: Optional[str] = FieldInfo(alias='fromEnvVar')


class OptionalValueFromEnvVar(BaseModel):
    value: Optional[str]
    from_env_var: Optional[str] = FieldInfo(alias='fromEnvVar')


class Config(BaseSettings):
    """Custom generator config options."""

    interface: InterfaceChoices = InterfaceChoices.asyncio
    partial_type_generator: Optional[Module]
    recursive_type_depth: int = FieldInfo(default=5)

    class Config(BaseSettings.Config):
        extra: Extra = Extra.forbid
        use_enum_values: bool = True
        env_prefix: str = 'prisma_py_config_'
        allow_population_by_field_name: bool = True

        @classmethod
        def customise_sources(
            cls,
            init_settings: SettingsSourceCallable,
            env_settings: SettingsSourceCallable,
            file_secret_settings: SettingsSourceCallable,
        ) -> Tuple[SettingsSourceCallable, ...]:
            # prioritise env settings over init settings
            return env_settings, init_settings, file_secret_settings

    @root_validator(pre=True)
    @classmethod
    def removed_http_option_validator(cls, values: Dict[str, Any]) -> Dict[str, Any]:
        http = values.get('http')
        if http is not None:
            if http in {'aiohttp', 'httpx-async'}:
                option = 'asyncio'
            elif http in {'requests', 'httpx-sync'}:
                option = 'sync'
            else:  # pragma: no cover
                # invalid http option, let pydantic handle the error
                return values

            raise ValueError(
                'The http option has been removed in favour of the interface option.\n'
                '  Please remove the http option from your Prisma schema and replace it with:\n'
                f'  interface = "{option}"'
            )
        return values

    @validator('partial_type_generator', pre=True, always=True, allow_reuse=True)
    @classmethod
    def partial_type_generator_converter(cls, value: Optional[str]) -> Optional[Module]:
        try:
            return Module(spec=value)
        except ValueError:
            if value is None:
                # no config value passed and the default location was not found
                return None
            raise

    @validator('recursive_type_depth', always=True, allow_reuse=True)
    @classmethod
    def recursive_type_depth_validator(cls, value: int) -> int:
        if value < -1 or value in {0, 1}:
            raise ValueError('Value must equal -1 or be greater than 1.')
        return value


class DMMF(BaseModel):
    datamodel: 'Datamodel'

    # TODO
    prisma_schema: Any = FieldInfo(alias='schema')


class Datamodel(BaseModel):
    enums: List['Enum']
    models: List['Model']


class Enum(BaseModel):
    name: str
    db_name: Optional[str] = FieldInfo(alias='dbName')
    values: List['EnumValue']


class EnumValue(BaseModel):
    name: str
    db_name: Optional[str] = FieldInfo(alias='dbName')


class Model(BaseModel):
    name: str
    is_embedded: bool = FieldInfo(alias='isEmbedded')
    db_name: Optional[str] = FieldInfo(alias='dbName')
    is_generated: bool = FieldInfo(alias='isGenerated')

    if TYPE_CHECKING:
        # pylint thinks all_fields is not an iterable
        all_fields: List['Field']
    else:
        all_fields: List['Field'] = FieldInfo(alias='fields')

    @property
    def related_models(self) -> Iterator['Model']:
        models = get_datamodel().models
        for field in self.relational_fields:
            for model in models:
                if field.type == model.name:
                    yield model

    @property
    def relational_fields(self) -> Iterator['Field']:
        for field in self.all_fields:
            if field.is_relational:
                yield field

    @property
    def scalar_fields(self) -> Iterator['Field']:
        for field in self.all_fields:
            if not field.is_relational:
                yield field

    @property
    def atomic_fields(self) -> Iterator['Field']:
        for field in self.all_fields:
            if field.type in ATOMIC_FIELD_TYPES:
                yield field

    @property
    def has_relational_fields(self) -> bool:
        try:
            next(self.relational_fields)
        except StopIteration:
            return False
        else:
            return True


class Field(BaseModel):
    name: str

    # TODO: switch to enums
    kind: str
    type: str

    is_id: bool = FieldInfo(alias='isId')
    is_list: bool = FieldInfo(alias='isList')
    is_unique: bool = FieldInfo(alias='isUnique')
    is_required: bool = FieldInfo(alias='isRequired')
    is_read_only: bool = FieldInfo(alias='isReadOnly')
    is_generated: bool = FieldInfo(alias='isGenerated')
    is_updated_at: bool = FieldInfo(alias='isUpdatedAt')

    default: Optional[Union['DefaultValue', str]]
    has_default_value: bool = FieldInfo(alias='hasDefaultValue')

    relation_name: Optional[str] = FieldInfo(alias='relationName')
    relation_on_delete: Optional[str] = FieldInfo(alias='relationOnDelete')
    relation_to_fields: Optional[List[str]] = FieldInfo(alias='relationToFields')
    relation_from_fields: Optional[List[str]] = FieldInfo(alias='relationFromFields')

    @root_validator
    @classmethod
    def scalar_type_validator(cls, values: Dict[str, Any]) -> Dict[str, Any]:
        kind = values.get('kind')
        type_ = values.get('type')

        if kind == 'scalar':
            if type_ is not None and type_ not in TYPE_MAPPING:
                raise ValueError(f'Unsupported scalar field type: {type_}')

        return values

    @validator('name')
    @classmethod
    def name_validator(cls, name: str) -> str:
        if getattr(BaseModel, name, None):
            raise ValueError(
                f'Field name "{name}" shadows a BaseModel attribute; '
                f'use a different field name with \'@map("{name}")\'.'
            )

        if iskeyword(name):
            raise ValueError(
                f'Field name "{name}" shadows a Python keyword; '
                f'use a different field name with \'@map("{name}")\'.'
            )

        return name

    # TODO: cache the properties
    @property
    def python_type(self) -> str:
        type_ = self._actual_python_type
        if self.is_list:
            return f'List[{type_}]'
        return type_

    @property
    def python_type_as_string(self) -> str:
        type_ = self._actual_python_type
        if self.is_list:
            type_ = type_.replace('\'', '\\\'')
            return f'\'List[{type_}]\''

        if not type_.startswith('\''):
            type_ = f'\'{type_}\''

        return type_

    @property
    def _actual_python_type(self) -> str:
        if self.kind == 'enum':
            return f'\'enums.{self.type}\''

        if self.kind == 'object':
            return f'\'models.{self.type}\''

        try:
            return TYPE_MAPPING[self.type]
        except KeyError as exc:
            # TODO: handle this better
            raise RuntimeError(
                f'Could not parse {self.name} due to unknown type: {self.type}',
            ) from exc

    @property
    def create_input_type(self) -> str:
        if self.kind != 'object':
            return self.python_type

        if self.is_list:
            return f'\'{self.type}CreateManyNestedWithoutRelationsInput\''

        return f'\'{self.type}CreateNestedWithoutRelationsInput\''

    @property
    def where_input_type(self) -> str:
        typ = self.type
        if typ in FILTER_TYPES:
            return f'Union[{self._actual_python_type}, \'types.{typ}Filter\']'
        if self.is_relational:
            if self.is_list:
                return f'\'{typ}ListRelationFilter\''
            return f'\'{typ}RelationFilter\''
        return self.python_type

    @property
    def relational_args_type(self) -> str:
        if self.is_list:
            return f'FindMany{self.type}Args'
        return f'{self.type}Args'

    @property
    def required_on_create(self) -> bool:
        return (
            self.is_required
            and not self.is_updated_at
            and not self.has_default_value
            and not self.relation_name
        )

    @property
    def is_optional(self) -> bool:
        return not (self.is_required and not self.relation_name)

    @property
    def is_relational(self) -> bool:
        return self.relation_name is not None

    @property
    def is_atomic(self) -> bool:
        return self.type in ATOMIC_FIELD_TYPES

    def maybe_optional(self, typ: str) -> str:
        """Wrap the given type string within `Optional` if applicable"""
        if self.is_required or self.is_relational:
            return typ
        return f'Optional[{typ}]'

    def get_update_input_type(self) -> str:
        if self.is_atomic:
            if self.is_list:
                raise NotImplementedError(
                    'Atomic updates for scalar list types not implemented yet.'
                )
            return f'Union[Atomic{self.type}Input, {self.python_type}]'

        if self.kind == 'object':
            if self.is_list:
                return f'\'{self.type}UpdateManyWithoutRelationsInput\''
            return f'\'{self.type}UpdateOneWithoutRelationsInput\''

        return self.python_type

    def get_relational_model(self) -> Optional['Model']:
        if not self.is_relational:
            return None

        name = self.type
        for model in get_datamodel().models:
            if model.name == name:
                return model
        return None


class DefaultValue(BaseModel):
    args: Any
    name: str


Enum.update_forward_refs()
DMMF.update_forward_refs()
Data.update_forward_refs()
Field.update_forward_refs()
Model.update_forward_refs()
Datamodel.update_forward_refs()
Generator.update_forward_refs()
Datasource.update_forward_refs()
