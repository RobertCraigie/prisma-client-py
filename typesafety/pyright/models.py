from prisma.models import Types, User


class MyUser(User):
    @property
    def hello(self):
        return f'Hello, {self.name}!'


async def create() -> None:
    # case: valid
    user = await User.prisma().create(
        data={
            'name': 'Robert',
        },
    )
    reveal_type(user)  # T: User

    user = await User.prisma().create(
        data={
            'name': 'Robert',
            'profile': {
                'create': {
                    'bio': 'hello',
                },
            },
        },
    )
    reveal_type(user)  # T: User

    type_ = await Types.prisma().create(data={})
    reveal_type(type_.bool)  # T: bool
    reveal_type(type_.string)  # T: str
    reveal_type(type_.bigint)  # T: int
    reveal_type(type_.integer)  # T: int
    reveal_type(type_.float_)  # T: float

    # case: subclassing
    user2 = await MyUser.prisma().create(
        data={
            'name': 'Robert',
        },
    )
    reveal_type(user2)  # T: MyUser
    reveal_type(user2.hello)  # T: str

    # case: invalid field
    await User.prisma().create(
        data={  # E: Argument of type "dict[str, str]" cannot be assigned to parameter "data" of type "UserCreateInput" in function "create"
            'foo': 'Robert',
        },
    )

    # case: invalid nested field
    await User.prisma().create(
        data={  # E: Argument of type "dict[str, str | dict[str, dict[str, str]]]" cannot be assigned to parameter "data" of type "UserCreateInput" in function "create"
            'name': 'Robert',
            'profile': {
                'create': {
                    'foo': 'bar',
                },
            },
        },
    )
