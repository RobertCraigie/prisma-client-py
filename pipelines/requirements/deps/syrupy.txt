syrupy==4.6.4
