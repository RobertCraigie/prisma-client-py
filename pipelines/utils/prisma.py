from __future__ import annotations

import nox


def generate(
    session: nox.Session,
    *,
    schema: str | None = 'tests/data/schema.prisma',
    clean: bool = True,
) -> None:
    if clean:
        session.run('python', '-m', 'prisma_cleanup')

    if schema:
        args = (f'--schema={schema}',)
    else:
        args = ()

    session.run('prisma', 'generate', *args)
