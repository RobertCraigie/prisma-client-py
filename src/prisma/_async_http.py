from typing import Any

import httpx

from ._types import Method
from .http_abstract import AbstractResponse, AbstractHTTP


__all__ = ('HTTP', 'Response', 'client')


class HTTP(AbstractHTTP[httpx.AsyncClient, httpx.Response]):
    # pylint: disable=invalid-overridden-method,attribute-defined-outside-init

    session: httpx.AsyncClient

    async def download(self, url: str, dest: str) -> None:
        # pyright thinks that stream() doesn't return a context manager
        # due to how httpx handles compatability imports
        # https://github.com/encode/httpx/discussions/1829
        async with self.session.stream(
            'GET', url, timeout=None
        ) as resp:  # pyright: reportGeneralTypeIssues=false
            with open(dest, 'wb') as fd:
                async for chunk in resp.aiter_bytes():
                    fd.write(chunk)

    async def request(self, method: Method, url: str, **kwargs: Any) -> 'Response':
        return Response(await self.session.request(method, url, **kwargs))

    def open(self) -> None:
        self.session = httpx.AsyncClient()

    async def close(self) -> None:
        if not self.closed:
            await self.session.aclose()

            # mypy doesn't like us assigning None as the type of
            # session is not optional, however the argument that
            # the setter takes is optional, so this is fine
            self.session = None  # type: ignore[assignment]


client: HTTP = HTTP()


class Response(AbstractResponse[httpx.Response]):
    # pylint: disable=invalid-overridden-method

    @property
    def status(self) -> int:
        return self.original.status_code

    async def json(self, **kwargs: Any) -> Any:
        return self.original.json(**kwargs)

    async def text(self, **kwargs: Any) -> Any:
        return self.original.content.decode(**kwargs)
