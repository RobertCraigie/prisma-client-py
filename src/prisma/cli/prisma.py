import os
import sys
import logging
import subprocess
from textwrap import indent
from typing import List, Optional, Dict

import click

from .. import binaries


log: logging.Logger = logging.getLogger(__name__)


def run(
    args: List[str],
    check: bool = False,
    env: Optional[Dict[str, str]] = None,
) -> int:
    directory = binaries.ensure_cached()
    path = directory.joinpath(binaries.PRISMA_CLI_NAME)
    if not path.exists():
        raise RuntimeError(
            f'The Prisma CLI is not downloaded, expected {path} to exist.'
        )

    log.debug('Using Prisma CLI at %s', path)
    log.debug('Running prisma command with args: %s', args)

    default_env = {
        **os.environ,
        'PRISMA_HIDE_UPDATE_MESSAGE': 'true',
        'PRISMA_CLI_QUERY_ENGINE_TYPE': 'binary',
    }
    env = {**default_env, **env} if env is not None else default_env
    # ensure the client uses our engine binaries
    for engine in binaries.ENGINES:
        env[engine.env] = str(engine.path.absolute())

    if args and args[0] == 'studio':
        click.echo(
            click.style(
                'ERROR: Prisma Studio does not work natively with Prisma Client Python',
                fg='red',
            ),
        )
        click.echo(
            '\nThere are two other possible ways to use Prisma Studio:\n'
        )
        click.echo(
            click.style('1. Download the Prisma Studio app\n', bold=True)
        )
        click.echo(
            indent(
                'Prisma Studio can be downloaded from: '
                + click.style(
                    'https://github.com/prisma/studio/releases', underline=True
                ),
                ' ' * 3,
            )
        )
        click.echo(
            click.style('\n2. Use the Node based Prisma CLI:\n', bold=True)
        )
        click.echo(
            indent(
                'If you have Node / NPX installed you can launch Prisma Studio by running the command: ',
                ' ' * 3,
            ),
            nl=False,
        )
        click.echo(click.style('npx prisma studio', bold=True))
        return 1

    process = subprocess.run(
        [str(path.absolute()), *args],
        env=env,
        check=check,
        stdout=sys.stdout,
        stderr=sys.stderr,
    )

    if args and args[0] in {'--help', '-h'}:
        prefix = ' '
        click.echo(click.style(prefix + 'Python Commands\n', bold=True))
        click.echo(
            prefix
            + 'For Prisma Client Python commands see '
            + click.style('prisma py --help', bold=True)
        )

    return process.returncode
