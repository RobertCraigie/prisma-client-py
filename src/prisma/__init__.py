# -*- coding: utf-8 -*-

__title__ = 'prisma'
__author__ = 'RobertCraigie'
__license__ = 'APACHE'
__copyright__ = 'Copyright 2020-2023 RobertCraigie'
__version__ = '0.13.0'

from typing import TYPE_CHECKING

from . import errors as errors
from .utils import setup_logging
from ._types import PrismaMethod as PrismaMethod
from ._config import config as config
from ._metrics import (
    Metric as Metric,
    Metrics as Metrics,
    MetricHistogram as MetricHistogram,
)
from .validator import *

# the import ordering is important here because
# we rely on the fact that `prisma/client.py` is the
# first piece of generated code that is loaded. This is
# especially useful for `python -m prisma_cleanup`
try:
    from .client import *  # noqa: I001, TID251
    from .fields import *  # noqa: TID251
    from . import (
        bases as bases,  # noqa: TID251
        types as types,  # noqa: TID251
        models as models,  # noqa: TID251
        partials as partials,  # noqa: TID251
    )
except ModuleNotFoundError:
    # code has not been generated yet
    # TODO: this could swallow unexpected errors

    # we only define this magic function when the client has not
    # been generated for performance reasons and we hide it from
    # type checkers so that they don't incidentally disable type
    # checking due to the dynamic nature of the `__getattr__`
    # magic dunder method.
    if not TYPE_CHECKING:

        def __getattr__(name: str):
            try:
                return globals()[name]
            except KeyError as err:
                # TODO: support checking for 'models' here too
                if name in {'Prisma', 'Client'}:
                    # TODO: remove this frame from the stack trace
                    raise RuntimeError(
                        "The Client hasn't been generated yet, "
                        'you must run `prisma generate` before you can use the client.\n'
                        'See https://prisma-client-py.readthedocs.io/en/stable/reference/troubleshooting/#client-has-not-been-generated-yet'
                    ) from None

                # leaves handling of this potential error to Python as per PEP 562
                raise AttributeError() from err


setup_logging()
