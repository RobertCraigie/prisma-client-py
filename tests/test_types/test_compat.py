from syrupy.assertion import SnapshotAssertion

from prisma.models import Types


def test_json_schema_compatible(snapshot: SnapshotAssertion) -> None:
    """Ensure a JSON Schema can be created for all types"""
    schema = Types.schema()
    assert schema == snapshot
