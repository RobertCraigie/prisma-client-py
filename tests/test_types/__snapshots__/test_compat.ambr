# name: test_json_schema_compatible
  dict({
    'description': 'Represents a Types record',
    'properties': dict({
      'bigint': dict({
        'title': 'Bigint',
        'type': 'integer',
      }),
      'bool_': dict({
        'title': 'Bool ',
        'type': 'boolean',
      }),
      'bytes': dict({
        'format': 'byte',
        'title': 'Bytes',
        'type': 'string',
      }),
      'datetime_': dict({
        'format': 'date-time',
        'title': 'Datetime ',
        'type': 'string',
      }),
      'decimal_': dict({
        'title': 'Decimal ',
        'type': 'number',
      }),
      'float_': dict({
        'title': 'Float ',
        'type': 'number',
      }),
      'id': dict({
        'title': 'Id',
        'type': 'integer',
      }),
      'integer': dict({
        'title': 'Integer',
        'type': 'integer',
      }),
      'optional_bigint': dict({
        'title': 'Optional Bigint',
        'type': 'integer',
      }),
      'optional_bool': dict({
        'title': 'Optional Bool',
        'type': 'boolean',
      }),
      'optional_bytes': dict({
        'format': 'byte',
        'title': 'Optional Bytes',
        'type': 'string',
      }),
      'optional_datetime': dict({
        'format': 'date-time',
        'title': 'Optional Datetime',
        'type': 'string',
      }),
      'optional_decimal': dict({
        'title': 'Optional Decimal',
        'type': 'number',
      }),
      'optional_float': dict({
        'title': 'Optional Float',
        'type': 'number',
      }),
      'optional_int': dict({
        'title': 'Optional Int',
        'type': 'integer',
      }),
      'optional_integer': dict({
        'title': 'Optional Integer',
        'type': 'integer',
      }),
      'optional_string': dict({
        'title': 'Optional String',
        'type': 'string',
      }),
      'string': dict({
        'title': 'String',
        'type': 'string',
      }),
    }),
    'required': list([
      'id',
      'bool_',
      'string',
      'bytes',
      'bigint',
      'integer',
      'float_',
      'datetime_',
      'decimal_',
    ]),
    'title': 'Types',
    'type': 'object',
  })
# ---
